### PR TITLE
Implements DreamMetaObjectIcon, Width(), Height()

### PR DESCRIPTION
**Pull request by ike709**
See #141 

~~This is some messy 3am coding, opening as a draft to make my life easier.~~

### CODE REVIEW

**Comment by ike709** [comment]
Probably good to review.

File: OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            public int Frame;
+            public byte? Moving;
+            public DMIParser.ParsedDMIDescription Description;
+
+            public DreamIcon(DreamResource rsc, DreamValue? state, DreamValue? dir, DreamValue? frame, DreamValue? moving)
+            {
+                Resource = rsc;
+                Description = DMIParser.ParseDMI(new MemoryStream(rsc.ResourceData));
+
+                // TODO confirm BYOND behavior of invalid types
```
Comment (ZeWaka) [inline OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs:25]: For the 'no name' empty icon state, it's an empty string. Default frames and dirs is 1. Moving default is null.
```
# BEGIN DMI
version = 4.0
	width = 32
	height = 192
state = "heavenlight"
	dirs = 1
	frames = 1
state = ""
	dirs = 1
	frames = 1
# END DMI
```
for 
image

Interesting tidbit: All trailing and preceding spaces in an icon state are invalid, and trimmed in DM.

File: OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            // These vars correspond to the args in icon/new() and the resulting /icon obj, not the actual DMI data
+            public string Icon;
+            public string? State; // Specific icon_state. Null is all states.
+            public AtomDirection? Direction; // Specific dir. Null is all dirs.
+            public byte? Frame; //1-indexed. Specific frame. Null is all frames.
+            public DreamIconMovingMode Moving;
+
+            public DreamIconObject(DreamResource rsc, DreamValue? state, DreamValue? dir, DreamValue? frame, DreamValue? moving)
+            {
+                if (rsc is ConsoleOutputResource)
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs:36]:
```suggestion
                if (Path.GetExtension(rsc.ResourcePath) != ".dmi")
```

File: OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            if (icon.TryGetValueAsDreamObjectOfType(DreamPath.Icon, out DreamObject copyFrom)) {
+                dreamIconObject = ObjectToDreamIcon[copyFrom];
+            } else if (icon.TryGetValueAsString(out string fileString))
+            {
+                var ext = Path.GetExtension(fileString);
+                switch (ext) // TODO implement other icon file types
+                {
+                    case ".dmi":
+                        dreamIconObject = new DreamIconObject(_rscMan.LoadResource(fileString), state, dir, frame, moving);
+                        dreamObject.SetVariableValue("icon", new DreamValue(fileString));
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs:112]: `/icon.icon` isn't the icon path. It seems to be some kind of internal structure, `usr << I.icon` leads to a "bad output value" error.
```suggestion
```

File: OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    case ".bmp":
+                        throw new NotImplementedException($"Unimplemented icon type '{ext}'");
+                    default:
+                        throw new Exception($"Invalid icon file {fileString}");
+                }
+
+            } else if (icon.TryGetValueAsDreamResource(out var rsc))
+            {
+                dreamIconObject = new DreamIconObject(rsc, state, dir, frame, moving);
+                dreamObject.SetVariableValue("icon", new DreamValue(rsc.ResourcePath));
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs:127]:
```suggestion
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            public DMIParser.ParsedDMIDescription Description; // TODO Eventually this should probably be removed in favor of just directly storing the data for the subset of the DMI that we actually care about
+
+            // These vars correspond to the args in icon/new() and the resulting /icon obj, not the actual DMI data
+            public string Icon;
+            public string? State; // Specific icon_state. Null is all states.
+            public AtomDirection? Direction; // Specific dir. Null is all dirs.
+            public byte? Frame; //1-indexed. Specific frame. Null is all frames.
+            public DreamIconMovingMode Moving;
+
+            public DreamIconObject(DreamResource rsc, DreamValue? state, DreamValue? dir, DreamValue? frame, DreamValue? moving)
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectIcon.cs:34]: No uses of this ever pass `null` values, these DreamValues don't need to be nullable.

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**